IBX-5385: Add option content-type to reindex command

papcio122 opened the pull request:
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-5385](https://issues.ibexa.co/browse/IBX-5385)
| **Type**                                   | bug/improvement
| **Target Ibexa version** | `v3.3`+
| **BC breaks**                          | no



#### Checklist:
- [ ] Provided PR description.
- [x] Tested the solution manually.
- [ ] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [ ] Asked for a review (ping `@ezsystems/engineering-team`).

papcio122 committed:
IBX-5385 add option content-type to reindex command

Steveb-p commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -128,23 +128,28 @@ protected function configure()
+            )->addOption(
+                'content-type',
+                null,
+                InputOption::VALUE_OPTIONAL,
```
If an option is passed, it's value should be required.
```suggestion
                InputOption::VALUE_REQUIRED,
```

An array of Content Types might also be considered:

```suggestion
                InputOption::VALUE_REQUIRED|InputOption::VALUE_IS_ARRAY,
```

Although this will require a few changes in code introduced here later on.

papcio122 committed:
Update eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php

Changed from VALUE_OPTIONAL to VALUE_REQUIRED for content-type option

Co-authored-by: Paweł Niedzielski <pawel.niedzielski@ibexa.co>

Steveb-p commented on eZ/Publish/SPI/Search/Content/IndexerGateway.php:
```diff
@@ -40,6 +40,18 @@ public function getContentInSubtree(string $locationPath, int $iterationCount):
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     *
+     * @return \Generator list of Content IDs for each iteration
+     */
+    public function getContentWithContentTypeIdentifier(string $contentTypeIdentifier, int $iterationCount): Generator;
```
For an interface a wider allowed type should be ok. What we're interested in is to be able to iterate over the results, so we can relax the constraints.

Ensure that what we're returning are integers.

```suggestion
     * @return iterable<int> list of Content IDs for each iteration
     */
    public function getContentWithContentTypeIdentifier(string $contentTypeIdentifier, int $iterationCount): iterable;
```

Steveb-p commented on eZ/Publish/SPI/Search/Content/IndexerGateway.php:
```diff
@@ -40,6 +40,18 @@ public function getContentInSubtree(string $locationPath, int $iterationCount):
     public function countContentInSubtree(string $locationPath): int;
 
+    /**
+     * @throws \Doctrine\DBAL\Exception
```
As far as the indexing is concerned, I'm not sure if we want to declare that this can throw this exception. The fact that we are using Doctrine is irrelevant for indexing (at least for the interface). @alongosz ?

alongosz replied:
> As far as the indexing is concerned, I'm not sure if we want to declare that this can throw this exception. The fact that we are using Doctrine is irrelevant for indexing (at least for the interface). @alongosz ?

You're right. On SPI level we can throw SPI or API exceptions.

Steveb-p commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -128,23 +128,28 @@ protected function configure()
+                'content-type',
+                null,
+                InputOption::VALUE_REQUIRED,
+                'Content type identifier to refresh (deleted/updated/added). Implies "no-purge", cannot be combined with "since", "subtree" or "content-ids"'
```
Side note, outside of the scope of this PR:
We are not actually checking that conflicting options are passed.

papcio122 committed:
Update eZ/Publish/SPI/Search/Content/IndexerGateway.php

Relaxed constraints for getContentWithContentTypeIdentifier in IndexerGateway interface.

Co-authored-by: Paweł Niedzielski <pawel.niedzielski@ibexa.co>

Steveb-p commented on eZ/Publish/Core/Search/Legacy/Content/IndexerGateway.php:
```diff
@@ -63,6 +63,22 @@ public function countContentInSubtree(string $locationPath): int
         return (int)$query->execute()->fetchOne();
     }
 
+    public function getContentWithContentTypeIdentifier(string $contentTypeIdentifier, int $iterationCount): Generator
```
The issue reported is relevant. Narrowing down of return type is available for PHP 7.4+, and this version of ezplatform-kernel supports PHP 7.3, which means you cannot use this feature.

```suggestion
    public function getContentWithContentTypeIdentifier(string $contentTypeIdentifier, int $iterationCount): iterable
```

Steveb-p commented on eZ/Publish/SPI/Search/Content/IndexerGateway.php:
```diff
@@ -40,6 +40,18 @@ public function getContentInSubtree(string $locationPath, int $iterationCount):
+     *
+     * @return iterable<int> list of Content IDs for each iteration
+     */
+    public function getContentWithContentTypeIdentifier(string $contentTypeIdentifier, int $iterationCount): iterable;
+
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function countContentWithContentTypeIdentifier(string $contentTypeIdentifier): int;
```
Are we actually ok with introducing new method into this class @alongosz? It is a contract.

alongosz replied:
> Are we actually ok with introducing new method into this class @alongosz? It is a contract.

No, this is strict SPI, any additions to contracts there (except for SPI\Persistence) are hard BC break.

OT: _Though, TBH I'm not sure what idea I had in mind when placing this in SPI. Maybe I thought it's gonna be implemented by each search engine, but AFAICS it should just fetch data related to indexing directly from Repository. Maybe the purpose was rather Legacy Storage strategy..._

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -128,23 +128,28 @@ protected function configure()
                 null,
                 InputOption::VALUE_OPTIONAL,
-                'Refresh changes since a time provided in any format understood by DateTime. Implies "no-purge", cannot be combined with "content-ids" or "subtree"'
+                'Refresh changes since a time provided in any format understood by DateTime. Implies "no-purge", cannot be combined with "content-ids", "subtree" or "content-type"'
```
Punctuation: use Oxford comma here and in the other related places please
```suggestion
                'Refresh changes since a time provided in any format understood by DateTime. Implies "no-purge", cannot be combined with "content-ids", "subtree", or "content-type"'
```

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -256,6 +261,10 @@ protected function indexIncrementally(
             $generator = $this->gateway->getContentInSubtree($location->pathString, $iterationCount);
             $purge = false;
+        } elseif ($contentType = $input->getOption('content-type')) {
+            $count = $this->gateway->countContentWithContentTypeIdentifier($contentType);
+            $generator = $this->gateway->getContentWithContentTypeIdentifier($contentType, $iterationCount);
+            $purge = false;
```
Given you can't add anything to that Gateway without breaking BC promise (see previous comment), I think better solution would be to either use Repository Filtering (`\eZ\Publish\API\Repository\ContentService::find`) or make a dedicated method in `\eZ\Publish\SPI\Persistence\Content\Handler`. The latter option would be more performant and require implementing cache layer (which is good).

alongosz reviewed (changes requested):
This improvement is not directly related to a bug fix, but it's rather a new feature making easier to reindex data after a bug fix, right? Moreover AFAICS related CS tickets are 4.x. We fix bugs and performance issues on the oldest supported version, but we don't add features to it.

As a compromise, I would accept this solution for 4.5 instead of 4.6.x-dev, if @webhdx agrees.

Review remarks:

papcio122 commented:
Closing this PR, moving changes to https://github.com/ibexa/core/pull/259 in order to merge this to 4.5
